### ISSUE TITLE
Selenium Grid maxSession ignored

### ISSUE THREAD

**Issue by adamlacoste** [SeleniumHQ/selenium#1694]
## Expected Behavior -
When I start a grid hub and node with maxSession 1, then make multiple simultaneous requests to the hub with webdriver, no more than one of those requests should be fulfilled at a time.

## Actual Behavior -
Making 5 simultaneous requests for a firefox session from the above setup results in 5 open firefox windows all processing at the same time.

## Steps to reproduce -
I'm building a node.js application using the [selenium-webdriver](https://www.npmjs.com/package/selenium-webdriver) and [selenium-server-standalone-jar](https://www.npmjs.com/package/selenium-server-standalone-jar) NPM modules.  The latter is just an include for selenium-server-standalone-2.52.0.jar.  The application follows this process:

### 1) Start a selenium grid hub running locally
    var jar = require("selenium-server-standalone-jar"),
        hub = new require("selenium-webdriver/remote").SeleniumServer(jar.path, {
            loopback: true,
            port:     4444,
            args:     [ "-role hub", "-hubConfig " + config_path ]
        });

That hub config file looks like this:

    {
        "host": null,
        "port": 4444,
        "newSessionWaitTimeout": -1,
        "throwOnCapabilityNotPresent": true,
        "nodePolling": 5000,
        "cleanUpCycle": 5000,
        "timeout": 300000,
        "browserTimeout": 0,
        "maxSession": 1
    }

### 2) Start a selenium grid node running locally
    var hub = new require("selenium-webdriver/remote").SeleniumServer(jar.path, {
            loopback: true,
            port:     5555,
            args:     [ "-role node", "-nodeConfig " + config_path ]
        });

The node config looks like this:

    {
        "capabilities": [
            {
                "browserName": "firefox",
                "maxInstances": 1,
                "seleniumProtocol": "WebDriver"
            }
        ],
        "configuration": {
            "maxSession": 1,
            "port": 5555,
            "host": "127.0.0.1",
            "register": true,
            "registerCycle": 5000,
            "hubPort": 4444,
            "hubHost": "127.0.0.1"
        }
    }

So at this point I have a hub running with maxSession=1 and a node running with maxSession=1 and maxInstances=1 (for Firefox). This isn't how the application will function long term, I'm just trying to get Selenium Grid to demonstrate that it CAN limit sessions.

### 3) Start 5 child processes, each with its own webdriver using the hub server
I use child_process.exec to split off multiple child processes, each executing cucumber.js.  Before each cucumber scenario, I create a new webdriver which uses the grid hub as its server, and after each scenario the webdriver is closed.

As mentioned above, I would expect only one of these child process webdrivers to be able to work at a time, but instead all do.  This presents a scaling problem for me; 5 browser windows is fine, but 50 probably would not be.  I need to be able to limit grid and the maxSession config parameter doesn't seem to be working.

**Comment by adamlacoste** [comment]
I have new information on this bug.  It seems to reside in the SeleniumServer construction; specifically, the "args" option array does not seem to be having any effect.

Some of the above behavior turned out to be the cause of a bug in my own logic where attempting to determine if a grid hub was active and using it if so; the 5 cucumber processes proceeded in parallel because they were detecting no active hub.  The symptom that revealed the root problem turned out to be that I couldn't reach http://localhost:4444/grid/console while running by test suite because the selenium servers were created without role args.

When I refactored my application to start the java processes using child_process.exec, everything seems to work more or less.  Still, this approach has problems because it creates a race condition between the grid startup and my cucumber processes; there is no readiness waiting as I would get from chaining off of the start() promise.

Is there any reason you can find in steps 1 and 2 above that my "args" list might not be working?  I'd really rather not resort to child_process as a final solution.

**Comment by lukeis** [comment]
Looks like the node package you're using is a separate project, please log your bug there:

https://github.com/adamhooper/selenium-server-standalone-jar

If there's any issues you're having with directly using the standalone server then we can further help you here.

**Comment by adamlacoste** [comment]
@lukeis The node package is just an npm wrapper for your jar file -- a convenient way to include it in the project, no more.  And as I mentioned above, I can get my code working if I use the jar file directly, so that's not where the problem is.  There's something wrong specifically with the args option or my use thereof.  If you can't see suggest something I'm doing wrong in setting the args option in the SeleniumServer constructor, I strongly object to this ticket being closed.

**Comment by lukeis** [comment]
Right now I can't see you showing us a bug in the standalone server and only a potential bug in the node package you're using, I don't see anything wrong necessarily with the args you are trying to use. But until shown otherwise without using the node wrapper, I see no bug in the selenium project itself

**Comment by adamlacoste** [comment]
Just to eliminate the possibility that that node module is at fault, I downloaded the jar from http://selenium-release.storage.googleapis.com/index.html?path=2.53/ and used this file when creating my SeleniumServers instead.  This had no effect.

However, I believe I have discovered the bug in my args arrays.  It expects parameter names and values to be separate items in the array, so `[ "-role hub", "-hubConfig " + config_path ]` doesn't work, but `[ "-role", "hub", "-hubConfig", config_path ]` probably does.

I say probably because the grid hub is timing out when I attempt to start() it.  (I've given it up to 60 seconds, with and without the json config specified.)  I can verify that the args are being honored however, by visiting grid/console and seeing a page with no nodes but a link to show the hub config.

Anyway, it's a new problem to troubleshoot so I guess it's fine to leave this issue closed.  If I can't figure out what's wrong now, I'll come back and make a new ticket.
